experimental/examples/token/Token_noTransferFrom.idr

```diff
@@ -49,42 +41,44 @@ getAccount address accounts = case lookup address accounts of
              True => let accountsStored = insert from (minus fromBalance tokens) (accounts storage) in
-                       Right (insert dest (destBalance + tokens) accountsStored)
+                       Right (record {accounts = (insert dest (destBalance + tokens) accountsStored)} storage)
 
 ||| createAccount transfers tokens from the owner to an address
 ||| @dest the address of the account to be created
 ||| @tokens the amount of tokens in the new created account
-total createAccount : (dest : Address) -> (tokens : Nat) -> (storage : Storage) -> Either Error (Either Error (SortedMap Address Account))
+total createAccount : (dest : Address) -> (tokens : Nat) -> (storage : Storage) -> Either Error Storage
 createAccount dest tokens storage =
-  let owner = owner storage in
+    let owner = owner storage in
       case owner == owner of --when sender can be detected, check sender == owner.
            False => Left FailedToAuthenticate
-           True => Right (performTransfer owner dest tokens storage)
-
-pvalues : SortedMap Address Account -> Type
-pvalues m = values m = []
-pvaluesempty : pvalues empty
-pvaluesempty = Refl
-
-{-
-psortedMape : (k : Address) -> (em : SortedMap Address Account) -> (em = empty) -> (lookup k em = Nothing)
-psortedMape _ Empty _ impossible
-psortedMape _ (M _ _) _ impossible
--}
-
---Prove the total supply is the sum of all account balances. TODO define owner account & balance and its relation to totalSupply.
-ptotalValues : Nat -> SortedMap Address Account -> Type
-ptotalSupplyToken : ptotalValues (totalSupply storage) (accounts storage)
-ptotalSupplyToken = ?rhs
-
-
-
---ptotalSupplyToken totalSupply (accounts )
---ptotalSupplyToken totalSupply allAccount prf = ?totalSupplyToken_rhs
+           True => performTransfer owner dest tokens storage
+-- **********End of contract without safety checks**********
+
+-- Below code is optional,
+-- running above functions (e.g., performTransfer) via provenAction adds safety listed in invariants.
+||| invariants checks whether certain invariants hold and returns True if they do.
+||| @oldStorage the storage before running the function
+||| @newStorage the storage after running the function
+invariants : (oldStorage : Storage) -> (newStorage : Storage) -> Bool
+invariants oldS newS =
+ (totalSupply oldS == totalSupply newS) && --totalSupply conserved
+ (owner oldS == owner newS) --the owner of the token contract is unchanged
+
+||| provenTotalSupplyAction checks that totalSupply is conserved from running the input function.
+||| @fn the input function (e.g., performTransfer) and all its input arguments except the storage
+||| @storage the storage input to fn
+provenAction : (fn : (Storage -> Either Error Storage)) -> (storage : Storage) -> Either Error Storage
+provenAction fn storage =
+ let result = fn storage in
+ case result of
+   Left _ => result
+   Right newStorage =>
+     if invariants storage newStorage then result
```

Comment:
This code will check the invariants at run-time, but we rather want to check them at compile-time (where possible) - try constructing a new `Storage` datatype which is indexed over the total supply, much like vectors are indexed over their lengths (see [here](http://docs.idris-lang.org/en/latest/tutorial/typesfuns.html#vectors)), and also enforces in the type that the sum of account balances is equal to the total supply index.

Reply:
Will do.